extern/sector-storage/stores/remote.go

```diff
@@ -415,4 +419,179 @@ func (r *Remote) FsStat(ctx context.Context, id ID) (fsutil.FsStat, error) {
+		}
+		log.Infof("partial file opened %s (+%d,%d)", path, offset, size)
+
+		// even though we have an unsealed file for the given sector, we still need to determine if we have the unsealed piece
+		// in the unsealed sector file. That is what `HasAllocated` checks for.
+		has, err := r.pfHandler.HasAllocated(pf, storiface.UnpaddedByteIndex(offset.Unpadded()), size.Unpadded())
+		if err != nil {
+			return nil, xerrors.Errorf("has allocated: %w", err)
+		}
+		log.Infof("partial file is allocated %s (+%d,%d)", path, offset, size)
+
+		if !has {
+			log.Infof("miner has unsealed file but not unseal piece, %s (+%d,%d)", path, offset, size)
+			if err := r.pfHandler.Close(pf); err != nil {
+				return nil, xerrors.Errorf("close partial file: %w", err)
+			}
+			return nil, nil
+		}
+
+		log.Infof("returning piece reader for local unsealed piece sector=%+v, (offset=%d, size=%d)", s.ID, offset, size)
+		return r.pfHandler.Reader(pf, storiface.PaddedByteIndex(offset), size)
+	}
+
+	// --- We don't have the unsealed sector file locally
+
+	// if we don't have the unsealed sector file locally, we'll first lookup the Miner Sector Store Index
+	// to determine which workers have the unsealed file and then query those workers to know
+	// if they have the unsealed piece in the unsealed sector file.
+	si, err := r.index.StorageFindSector(ctx, s.ID, ft, 0, false)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(si) == 0 {
+		return nil, xerrors.Errorf("failed to read sector %v from remote(%d): %w", s, ft, storiface.ErrSectorNotFound)
+	}
+
+	sort.Slice(si, func(i, j int) bool {
+		return si[i].Weight > si[j].Weight
+	})
+
+	for _, info := range si {
+		for _, url := range info.URLs {
+			// checkAllocated makes a JSON RPC query to a remote worker to determine if it has
+			// unsealed piece in their unsealed sector file.
+			ok, err := r.checkAllocated(ctx, url, s.ProofType, offset, size)
+			if err != nil {
+				log.Warnw("check if remote has piece", "url", url, "error", err)
+				continue
+			}
+			if !ok {
+				continue
+			}
+
+			// readRemote fetches a reader that we can use to read the unsealed piece from the remote worker.
+			// It uses a ranged HTTP query to ensure we ONLY read the unsealed piece and not the entire unsealed file.
+			rd, err := r.readRemote(ctx, url, offset, size)
+			if err != nil {
+				log.Warnw("reading from remote", "url", url, "error", err)
+				continue
+			}
+			log.Infof("Read remote %s (+%d,%d)", url, offset, size)
+			return rd, nil
+		}
+	}
+
+	// we couldn't find a unsealed file with the unsealed piece, will return a nil reader.
+	log.Debugf("returning nil reader, did not find unsealed piece for %+v (+%d,%d)", s, offset, size)
+	return nil, nil
+}
+
+func (r *Remote) Reserve(ctx context.Context, sid storage.SectorRef, ft storiface.SectorFileType, storageIDs storiface.SectorPaths, overheadTab map[storiface.SectorFileType]int) (func(), error) {
+	panic("not implemented")
+}
```

Comment:
Where is this used?

Reply:
@magik6k 

For unit testing, I needed the `Store` interface to have this method so I could replace a concrete `LocalStore` implementation with a `Store` interface to then mock it out.

Reply:
Can we at least not panic here, just in case this gets called somewhere by accident?

Reply:
done.